src/sentry/snuba/discover.py

```diff
@@ -0,0 +1,230 @@
+from __future__ import absolute_import
+
+import six
+
+from copy import deepcopy
+from sentry.api.event_search import get_filter, resolve_field_list
+from sentry.utils.snuba import (
+    Dataset,
+    DISCOVER_COLUMN_MAP,
+    QUOTED_LITERAL_RE,
+    get_function_index,
+    raw_query,
+)
+
+ALLOWED_INTERNALS = set(["project_id", "event_id"])
```

Comment:
Why are these internal syntax allowed in the public API?

Reply:
This is a temporary measure, as the aliased field macros reference these columns. I could try changing the aliased field definitions but it will churn code in the other discover implementation.

Reply:
I think we should try to be very careful about where we are dealing with public names and where these are being converted to internal names. That should happen at the same point for all the parts of the query.

If I understand correctly right now we have `event_search` returning public names almost everywhere except for in these aliased field definitions which are still using the internal names.

Reply:
These fields have their internal names used in the automatically appended aggregations for `projectid`. I'll see if those references can be updated to use the public names without impacting the existing discover implementation.

Reply:
Ah - is that because none of Discover is actually using this code yet? Related to this, what is the rollout plan around this new Discover dataset API?

Reply:
Yes, none of the endpoints are using this code yet. My plan was:

1. Get a rough implementation of the query method. (This pull request)
2. Update one endpoint at a time to use the new query method.
3. Get a rough implementation of the timeseries high level function built.
4. Update the timeseries endpoint to use the new method.
5. Get the pagination method built in this module.
6. Update event details to use the new pagination function.
7. Update discover1 to use the new discover methods.
8. Remove `transform_aliases_and_query` once it becomes unused.
9. Pare down dataset_query and related helper function usage to only resolve aliases for the events + groups dataset as Chris is working in that area still.

I thought breaking up the first pass implementation of the new methods from the endpoint cut overs would make for more manageable pull requests as not everything was changing all at once.

Reply:
Plan sounds good, I saw that ALLOWED_INTERNALS is gone now - is this causing the issues you mentioned it might in the existing implementation?

Reply:
Not that I was able observe.